### PR TITLE
Add support for temple, multiple creature meshes and cam files

### PR DESCRIPTION
**Pull request by raffclar**
## GENERAL
* We've got entities, we've got components but we haven't developed any systems yet. This PR adds the interface for all systems going forward.

* Added camera movement smoothing using a cheap trick. Makes camera movement feel more natural

## Meshes
* You can spawn meshes using the mesh viewer
* Alpha transprency threshold has been adjusted to make trees look better at the cost of breaking other textures in the interior temple meshes.
* Changed BGFX max bones count from 64 to 128 as some of the creatures exceeded 64.
* Added a bunch of out-of-bounds checks relating to meshes, textures and submeshes.
* Added basic support for Black & White: Creature Isle meshes by adding a workaround for DXT5 textures. Openblack won't segfault immediately.
* Added support for LZW compressed meshes - these compressed meshes are used for the temple interiors and others bits.
* We can now read the name data from meshes

## CAM files
* We now have basic support for camera path files (.CAM). A .CAM file contains a list of structs where each struct defines a 3D position and a rotation. The number of structs in the list is defined in the file header along with a movement speed multiplier.
* To start off with a new ECS system has been added that controls the player's camera position and speed. The camera follows a given camera path as defined in our CAM files.

## TEMPLE
* Added support for entering and leaving the temple interior using a debug dialog
* Temple interior rooms are ECS entities that are assigned mesh components
* Support for camera path transitioning has been added. The initial 360 degree spin upon entering the temple is played

## CREATURE
* Added support for loading all creatures meshes along with their variants (good, evil). Creatures are now spawned via LHScript.

### CODE REVIEW

File: components/l3d/src/L3DFile.cpp [hunk 1]
```diff
@@ -473,6 +473,46 @@ void L3DFile::ReadFile(std::istream& stream)
 		}
 	}
 
+	// Get additional data. Strictly in this order
+	// Footprint data
+	uint32_t additionalDataOffset = 0;
+	uint32_t footprintDataSize = 0;
+	if (static_cast<uint32_t>(_header.flags) & static_cast<uint32_t>(L3DMeshFlags::ContainsLandscapeFeature))
+	{
+		stream.seekg(0x48, stream.beg);
```
Comment (bwrsandman) [inline components/l3d/src/L3DFile.cpp:482]: Where does this magic number come from?
Could you make a const with its derivation available? Either a bunch of sizeofs or offsetofs would make it easier to understand for others.

File: components/pack/src/PackFile.cpp [hunk 2]
```diff
@@ -282,9 +282,16 @@ void PackFile::ExtractTexturesFromBlock()
...
 		ddsStream.read(reinterpret_cast<char*>(&ddsHeader), sizeof(DdsHeader));
 
-		// TODO(bwrsandman) the extra sizeof(uint32_t) is unaccounted for
-		if (header.ddsSize - sizeof(ddsHeader) - sizeof(uint32_t) != ddsHeader.pitchOrLinearSize)
+		// Workaround for non-conforming DXT5 textures in Creature Isle
+		if (ddsHeader.pitchOrLinearSize == 0 && ddsHeader.format.fourCC == std::string("DXT5"))
 		{
+			auto blockSize = 16;
+			ddsHeader.pitchOrLinearSize = ((ddsHeader.width + 3) / 4) * ((ddsHeader.height + 3) / 4) * blockSize;
+			;
```
Comment (bwrsandman) [inline components/pack/src/PackFile.cpp:290]: dangling semi

File: src/3D/Camera.cpp [hunk 3]
```diff
@@ -8,10 +8,14 @@
  *****************************************************************************/
 
 #include "Camera.h"
-#include <3D/LandIsland.h>
-#include <Game.h>
+#include "Game.h"
+#include "3D/LandIsland.h"
```
Comment (bwrsandman) [inline src/3D/Camera.cpp:12]: If you're going to use relative imports, you should use them based on the current directory:
```c++
#include "../Game.h"
#include "../Entities/Registry.h"
#include "LandIsland.h"
```

File: src/3D/Camera.cpp [hunk 4]
```diff
@@ -183,9 +199,23 @@ void Camera::handleMouseInput(const SDL_Event& e)
 	}
 }
 
+#define BV(v) (btVector3((v).x, (v).y, (v).z))
+#define AIR_DENSITY (1.2f)
```
Comment (bwrsandman) [inline src/3D/Camera.cpp:203]: Use a static const, preferably as part of a namespace.

File: src/3D/Camera.cpp [hunk 5]
```diff
@@ -183,9 +199,23 @@ void Camera::handleMouseInput(const SDL_Event& e)
 	}
 }
 
+#define BV(v) (btVector3((v).x, (v).y, (v).z))
```
Comment (bwrsandman) [inline src/3D/Camera.cpp:202]: Are we using Bullet now?

File: src/3D/Camera.cpp [hunk 6]
```diff
@@ -183,9 +199,23 @@ void Camera::handleMouseInput(const SDL_Event& e)
...
+#define AIR_DENSITY (1.2f)
+
 void Camera::Update(std::chrono::microseconds dt)
 {
-	_position += _velocity * (_movementSpeed * dt.count());
+	auto maxSpeed = 0.005f;
+	auto accelFactor = 0.001f;
+	auto ghettoAirResistance = .9f;
```
Comment (bwrsandman) [inline src/3D/Camera.cpp:209]: Odd variable name???

File: src/3D/Camera.h [hunk 7]
```diff
@@ -71,16 +81,20 @@ class Camera
...
 	glm::vec3 _rotation;
 	glm::vec3 _dv;
 
 	glm::mat4 _projectionMatrix;
 
 	glm::vec3 _velocity;
-	float _movementSpeed;
+	glm::vec3 _movementSpeed;
+	float _maxMovementSpeed = .0005f;
+	glm::vec3 _acceleration = glm::vec3(0.0f);
```
Comment (bwrsandman) [inline src/3D/Camera.h:97]: Can you put these definitions in the constructor instead?

File: src/3D/Camera.h [hunk 8]
```diff
@@ -30,8 +34,12 @@ class Camera
...
-	    , _movementSpeed(0.0005f)
+	    , _movementSpeed(0.0000f)
+	    , _maxMovementSpeed(0.0005f)
 	    , _freeLookSensitivity(1.0f)
+	    , _focused(false)
+	    , _focusedPosition(0.0f)
+	    , _focusedPositionRadius(0.0f)
 	{
 	}
 	Camera()
```
Comment (bwrsandman) [inline src/3D/Camera.h:45]: You should move all function definitions that aren't very basic to an implementation file.
It'll speed up compilation and make it more clean to read and maintain.

File: src/3D/Camera.h [hunk 9]
```diff
@@ -47,6 +55,8 @@ class Camera
 
 	[[nodiscard]] glm::vec3 GetPosition() const { return _position; }
 	[[nodiscard]] glm::vec3 GetRotation() const { return glm::degrees(_rotation); }
+	[[nodiscard]] glm::vec3 GetMovementSpeed() const { return _movementSpeed; }
```
Comment (bwrsandman) [inline src/3D/Camera.h:58]: Speed + direction become a velocity.
```
GetVelocity
_velocity
```

File: src/3D/CameraPath.cpp [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+	file.Open(path.string());
+	auto filePoints = file.GetPoints();
+
+	for (auto& filePoint : filePoints)
+	{
+		CameraPoint node;
+		node.start = _points.size() == 0;
+		node.position = glm::vec3(filePoint.x, filePoint.y, filePoint.z);
+		node.rotation = glm::normalize(node.position + glm::vec3(filePoint.pitch, filePoint.yaw, filePoint.roll));
+		_points.emplace_back(node);
```
Comment (bwrsandman) [inline src/3D/CameraPath.cpp:32]: C++ question here: Does this do a move or copy? Seems like there will be a copy done here no matter what.
Maybe use the reference returned by emplace instead?
```c++
CameraPoint& node = _points.emplace_back();
```

File: src/3D/CameraPath.h [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+	bool LoadFromFile(const fs::path& path);
+	const std::vector<CameraPoint>& GetPoints() const { return _points; }
+	void CreatePathEntities(glm::vec3 position);
+	void CreatePathStartEntity(glm::vec3 position);
+	void CreatePathEndEntity(glm::vec3 position);
+	glm::vec3 GetPathStartPosition() const { return _points.front().position; }
+	void RemoveFromRegistry();
+private:
+	std::vector<CameraPoint> _points;
+	entt::entity _start;
```
Comment (bwrsandman) [inline src/3D/CameraPath.h:43]: Can you explain why we're using a linked list instead of a vector?

File: src/3D/MeshPack.cpp [hunk 12]
```diff
@@ -10,13 +10,17 @@
...
 #include "AllMeshes.h"
 #include "Common/FileSystem.h"
 #include "Common/MemoryStream.h"
 #include "Entities/Components/Hand.h"
 #include "Game.h"
 #include "Graphics/Texture2D.h"
 
+extern "C" {
+#include "miniz.c"
+}
```
Comment (bwrsandman) [inline src/3D/MeshPack.cpp:23]: Why include a c file here?

**Review by bwrsandman — CHANGES_REQUESTED** [review]
Lots of really nice things here.

Unfortunately the change is quite big and I have a few questions and changes.

Would you mind splitting this PR up in multiple smaller PRs based on logical changes with descriptions. It would make it easier for us to review.
Right now, I don't know what I'm reviewing: Camera pathing, improvements to meshes, support for .zzz files or others!

I promise, if you do so, I will review them quickly ;)

File: apps/l3dtool/l3dtool.cpp [hunk 13]
```diff
@@ -569,7 +569,7 @@ int WriteFile(const Arguments::Write& args)
 		L3DSubmeshHeader submesh;
 		submesh.flags.lod = 0;
 		submesh.flags.status = 0;
-		submesh.flags.unknown1 = 0b0101;
+		submesh.flags.unknown1 = 0b01;
```
[the remark this answers is not in the document]
Comment (raffclar) [reply apps/l3dtool/l3dtool.cpp]: Sadly no. I have undone the change

File: components/l3d/include/L3DFile.h [hunk 14]
```diff
@@ -129,10 +144,29 @@ struct L3DSubmeshHeader
 	{
 		uint32_t lod : 3;
 		uint32_t status : 6;
-		uint32_t unknown1 : 3; // always 0b0101
+		uint32_t unknown1 : 1; // always 0b0101
+		uint32_t unknown2 : 1; // always 0b0101
+		uint32_t unknown3 : 1; // always 0b0101
```
[the remark this answers is not in the document]
Comment (raffclar) [reply components/l3d/include/L3DFile.h]: Was part of an attempt to explore these mesh flags in search of transparency related flags

[the remark this answers is not in the document]
Comment (raffclar) [reply components/l3d/include/L3DFile.h]: Change undone

File: components/l3d/include/L3DFile.h [hunk 15]
```diff
@@ -212,6 +246,15 @@ struct L3DBlend
...
 
+struct UV2Data
+{
+
+};
+
+struct FootprintData
+{
+};
+
```
[the remark this answers is not in the document]
Comment (raffclar) [reply components/l3d/include/L3DFile.h]: Yep. Only NameData was explored. UV2 and FootPrint needs work. Removed the empty structs for now

File: src/3D/Camera.cpp (hunk 4 above)
Comment (raffclar) [reply src/3D/Camera.cpp]: Removed these macros. This was part of an attempt to model camera movement physics in a more realistic manner

File: src/3D/Camera.cpp (hunk 6 above)
Comment (raffclar) [reply src/3D/Camera.cpp:202]: Renamed

File: src/3D/Camera.cpp (hunk 5 above)
Comment (raffclar) [reply src/3D/Camera.cpp]: I want to ;_;

File: src/3D/Camera.h (hunk 8 above)
Comment (raffclar) [reply src/3D/Camera.h]: Removed camera focusing until temple movement is better understood

File: src/3D/CameraPath.cpp (hunk 10 above)
Comment (raffclar) [reply src/3D/CameraPath.cpp:29]: Good idea

File: src/3D/Camera.h (hunk 9 above)
Comment (raffclar) [reply src/3D/Camera.h:51]: Good catch. Thanks for spotting this. I wasn't using `_velocity` at all

File: src/3D/Camera.h (hunk 7 above)
Comment (raffclar) [reply src/3D/Camera.h]: Done

File: src/3D/MeshPack.cpp (hunk 12 above)
Comment (raffclar) [reply src/3D/MeshPack.cpp:20]: Needed for ".zzz" compressed files. We could create new component for LZW compression and wrap this C library. Also vcpkg supports miniz https://github.com/microsoft/vcpkg/pull/5452 so we could tackle that as well. Thoughts?

File: components/cam/include/CamFile.h [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+	uint32_t unk1; // No perceptible difference when altered. Some values cause the game to crash
+	uint32_t movementSpeed;  // The larger this integer, the slower the movement speed
+	uint32_t numberOfPoints; // How many path vertices there are in the file
+};
+
+struct CamPoint
+{
+	float x;
+	float y;
+	float z;
```
[the remark this answers is not in the document]
Comment (raffclar) [reply components/cam/include/CamFile.h]: Used glm::vec3

File: src/3D/MeshPack.cpp (hunk 12 above)
Comment (raffclar) [reply src/3D/MeshPack.cpp:20]: Thoughts @bwrsandman ?

**Comment by raffclar** [comment]
Splitting into multiple PRs

**Pull request closed without merging**